proxy/src/telemetry/metrics/latency.rs

```diff
@@ -0,0 +1,267 @@
+#![deny(missing_docs)]
+use control::pb::telemetry::LatencyBucket;
+
+use std::{ops, slice, u32};
+use std::default::Default;
+use std::time::Duration;
+
+/// The number of buckets in a  latency histogram.
+pub const NUM_BUCKETS: usize = 26;
+
+/// A series of latency values and counts.
+#[derive(Debug)]
+pub struct Histogram([Bucket<Latency>; NUM_BUCKETS]);
+
+/// A latency in tenths of a millisecond.
+#[derive(Debug, Default, Eq, PartialEq, Ord, PartialOrd, Copy, Clone, Hash)]
+pub struct Latency(u32);
+
+/// A "leaf" bucket containing a single counter.
+///
+/// # Type Parameters:
+/// - `T`: the type of observations counted by this bucket.
+/// - `C`: the type used for counting observed values.
+#[derive(Debug, Clone)]
+pub struct Bucket<T, C = u32> {
+
+    /// The upper bound of the range of values in this bucket.
+    max: T,
+
+    /// Count of observations falling into this bucket.
+    count: C,
+
+}
+
+
+impl<T> Bucket<T> {
+
+    // Construct a new `LeafBucket` with the upper bound of `max`.
+    fn new<I: Into<T>>(max: I) -> Self {
+        Bucket {
+            max: max.into(),
+            count: 0
+        }
+    }
+
+}
+
+impl<T> Bucket<T>
+where
+    T: PartialOrd
+{
+
+    fn contains(&self, t: &T) -> bool {
+        t <= &self.max
```

Comment:
or an OrderMap of bounds -> counts

Reply:
I considered having the `Bucket` store a _range_ rather than just the `max`, and test if the observed value is in the bucket's range, but this seemed unnecessary if we know the buckets will be stored in increasing order. If you think having an explicit lower bound on the buckets matters, though, I could do it that way?